### PR TITLE
Proposal: Try to separate out the serialization soup.

### PR DESCRIPTION
**Pull request by AlCutter**
Suggestions on how to separate this out welcome!

### CODE REVIEW

**Comment by Martin2112** [comment]
Can you upload this to rietveld? I think it will be easier to review it there.

**Comment by Martin2112** [comment]
Test failure was a flake. Going to merge.

**Comment by AlCutter** [comment]
Here you go: https://codereview.appspot.com/282720043

**Comment by Martin2112** [comment]
Thanks. Doesn't look much less scary on there but should be easier to
manage.

Martin

**Comment by pphaneuf** [comment]
I recommend you hold off before reviewing this, I bring relief, in the form of a large cleaver. Stand by...

**Comment by Martin2112** [comment]
I've started reviewing but I'll wait to see what happens after the cleaver
impacts.

**Comment by pphaneuf** [comment]
@Martin2112, you should prioritize #1076 over this one, as this here will churn a bit until the others have all landed. Then, this one will be easier. :smiley:

**Comment by AlCutter** [comment]
#1076 is in; I've rebased this on top.

**Comment by AlCutter** [comment]
Pierre's got his cleaver out again in #1077, will rebase again after that.

**Comment by Martin2112** [comment]
There is no stopping him!

**Comment by pphaneuf** [comment]
This was 800 and 600 lines taken out of this complicated review, with two not-complicated PRs. I really hope you guys are **ALL** going to keep a cleaver (or an axe?) by your keyboards, in the future, for all our sakes... :wink:

It didn't take me that long to do, and a lot of the time it took was to reverse-engineer what made sense in one PR or the other (which might have taken @AlCutter less time than it took me).

**Comment by AlCutter** [comment]
Which is why it was sent out as a proposal with the description asking for suggestions on ways to split it out...  Nonetheless, I do appreciate you taking the time to actually do that splitting out.

**Comment by AlCutter** [comment]
#1077 is in, rebased.

File: cpp/proto/xjson_serializer.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#include <glog/logging.h>
+#include <google/protobuf/repeated_field.h>
+#include <stdint.h>
+#include <string>
+
+#include "base/macros.h"
+#include "proto/ct.pb.h"
+
+
+const size_t kMaxJsonLength = (1 << 24) - 1;
```
Comment (pphaneuf) [inline cpp/proto/xjson_serializer.h]: This is only used inside `xjson_serializer.cc`? Hide it there?

Comment (AlCutter) [inline cpp/proto/xjson_serializer.h]: Done.

File: cpp/proto/cert_serializer.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#include <stdint.h>
+#include <string>
+
+#include "base/macros.h"
+#include "proto/ct.pb.h"
+#include "proto/serializer.h"
+
+
+const size_t kMaxCertificateLength = (1 << 24) - 1;
+const size_t kMaxCertificateChainLength = (1 << 24) - 1;
```
Comment (pphaneuf) [inline cpp/proto/cert_serializer.h]: Hide these two as well?

File: cpp/proto/serializer.cc [hunk 3]
```diff
@@ -750,6 +364,9 @@ SerializeResult Serializer::SerializeDigitallySigned(
   return SerializeResult::OK;
 }
 
+#if 0
```
Comment (pphaneuf) [inline cpp/proto/serializer.cc]: You're going to actually delete these, right?

File: cpp/proto/cert_serializer.h (hunk 2 above)
Comment (AlCutter) [inline cpp/proto/cert_serializer.h]: Done

File: cpp/proto/serializer.cc (hunk 3 above)
Comment (AlCutter) [inline cpp/proto/serializer.cc]: Odd, I'm pretty sure I did once already - gone again.

File: cpp/proto/serializer_test.cc [hunk 4]
```diff
@@ -203,6 +206,7 @@ const char kDefaultCertSCTSignedHexStringV2Extensions[] =
     // extension 2 data "thx"
     "746878";
 
+#if 0
```
Comment (pphaneuf) [inline cpp/proto/serializer_test.cc]: And this?

File: cpp/proto/serializer.h [hunk 5]
```diff
@@ -307,64 +255,37 @@ class Serializer {
...
+      serialize_sct_merkle_leaf_;
+  static std::function<SerializeResult(uint64_t timestamp, int64_t tree_size,
+                                       const std::string& root_hash,
+                                       std::string* result)>
+      serialize_sth_sig_input_v1_;
+  static std::function<SerializeResult(
+      uint64_t timestamp, int64_t tree_size, const std::string& root_hash,
+      const repeated_sth_extension& sth_extension, const std::string& log_id,
+      std::string* result)>
+      serialize_sth_sig_input_v2_;
```
Comment (pphaneuf) [inline cpp/proto/serializer.h]: These are a fair bit of innards showing, you could remove those from here, I think?

File: cpp/proto/serializer.h [hunk 6]
```diff
@@ -405,14 +326,12 @@ class Deserializer {
...
   Deserializer() = delete;
 
-  // FIXME(ekasper): for simplicity these reject if the list has empty
-  // elements (all our use cases are like this) but they should take in
-  // an arbitrary min bound instead.
-  static DeserializeResult DeserializeList(const std::string& in,
-                                           size_t max_total_length,
-                                           size_t max_elem_length,
-                                           repeated_string* out);
+  static std::function<DeserializeResult(TLSDeserializer* d,
```
Comment (pphaneuf) [inline cpp/proto/serializer.h]: Same thing for this one, I don't think it needs to be here.

File: cpp/proto/serializer.h (hunk 5 above)
Comment (AlCutter) [inline cpp/proto/serializer.h]: removed.

File: cpp/proto/serializer.h (hunk 6 above)
Comment (AlCutter) [inline cpp/proto/serializer.h]: removed

File: cpp/proto/serializer_test.cc (hunk 4 above)
Comment (AlCutter) [inline cpp/proto/serializer_test.cc]: removed

**Pull request merged**